src/ring_theory/finiteness.lean

```diff
@@ -550,3 +551,73 @@ ring_hom.finite_presentation.of_finite_type
 end finite_presentation
 
 end alg_hom
+
+section monoid_algebra
+
+namespace add_monoid_algebra
+
+open algebra add_submonoid
+
+variables {R : Type*} {M : Type*} [comm_ring R] [add_comm_monoid M]
+
+lemma surjective_of_closure {S : set M} (hS : closure S = ⊤) :
+  function.surjective (mv_polynomial.aeval (λ (s : S), of R M (multiplicative.of_add ↑s)) :
+  mv_polynomial S R → add_monoid_algebra R M) :=
+begin
+  refine λ f, finsupp.induction_linear f _ _ (λ m r, _),
+  { exact ⟨0, alg_hom.map_zero _⟩ },
+  { rintro f g ⟨P, rfl⟩ ⟨Q, rfl⟩,
+    exact ⟨P + Q, alg_hom.map_add _ _ _⟩, },
+  { suffices : ∃ (P : mv_polynomial S R), (mv_polynomial.aeval
+      (λ (s : S), of R M (multiplicative.of_add ↑s))) P = finsupp.single m 1,
+    { obtain ⟨P, hP⟩ := this,
+      refine ⟨r • P, by simp only [hP, alg_hom.map_smul, mul_one, finsupp.smul_single']⟩ },
+    have : m ∈ closure S := hS.symm ▸ mem_top _,
+    refine closure_induction this (λ m hm, _) _ _,
+    { exact ⟨mv_polynomial.X ⟨m, hm⟩, mv_polynomial.aeval_X _ _⟩ },
+    { exact ⟨1, alg_hom.map_one _⟩ },
+    { rintro m₁ m₂ ⟨P₁, hP₁⟩ ⟨P₂, hP₂⟩,
+      exact ⟨P₁ * P₂, by rw [alg_hom.map_mul, hP₁, hP₂, single_mul_single, one_mul]⟩ } }
```

Comment:
If you're looking for further refactor... There's definitely an `add_monoid_algebra.induction` hiding here that extends `finsupp.induction_linear` with statements about `1`, `*`, and `•`

Reply:
I've added `monoid_algebra.induction` (and `add_monoid_algebra.induction`) and used them in the proof of `surjective_of_closure`. I don't think it's enough to assume the result for `1`, so the statement is not so different from `finsupp.induction_linear`, but it seems still useful.